src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp

```diff
@@ -0,0 +1,145 @@
+                   distance(defender_assignment.target, block_kick_point) <=
+                       TOO_CLOSE_THRESHOLD_METERS)
+            {
+                assignments_skipped.push(defender_assignment);
+                defender_assignment = assignments.at(i + assignments_skipped.size());
+            }
+        }
+        else
+        {
+            // If we have more tactics to set than determined defender assignments:
+            // first assign remaining defenders to the previously skipped defender
+            // assignments otherwise, assign a defender to stand between the kicker and
+            // net if spacing permits else default to defender with the highest coverage
+            // rating
+            if (!assignments_skipped.empty())
+            {
+                defender_assignment = assignments_skipped.front();
+                assignments_skipped.pop();
+            }
+            else if (tactics_to_return[0].size() < 2 &&
+                     distance(event.common.world_ptr->field().friendlyGoalCenter(),
+                              block_kick_point) >=
+                         event.common.world_ptr->field().totalYLength() / 2)
+            {
+                auto mid_zone_defender = std::make_shared<PassDefenderTactic>();
+                Point mid_point        = Point(
+                    (event.common.world_ptr->ball().position().toVector() +
+                     event.common.world_ptr->field().friendlyGoalCenter().toVector()) /
+                    2);
+                mid_zone_defender->updateControlParams(mid_point);
+                tactics_to_return[0].push_back(mid_zone_defender);
+            }
+            else
+            {
+                defender_assignment = assignments.front();
+            }
+        }
+
+        if (defender_assignment.type == CREASE_DEFENDER)
+        {
+            crease_defender_assignments.emplace_back(defender_assignment);
+
+            // If we have at least two available defenders, two defenders should
+            // be assigned to the highest scoring crease defender assignment to better
+            // block the shot cone of the most threatening enemy
+            if (i == 0 && num_defenders >= 2)
+            {
+                crease_defender_assignments.emplace_back(defender_assignment);
+                i++;
+            }
+        }
+        else
+        {
+            pass_defender_assignments.emplace_back(defender_assignment);
+        }
+    }
+
+    // Reset tactics if the number of crease defenders or pass defenders
+    // we intend to assign has changed
+    DefensePlayBase::setUpCreaseDefenders(static_cast<unsigned int>(crease_defender_assignments.size()));
+    DefensePlayBase::setUpPassDefenders(static_cast<unsigned int>(pass_defender_assignments.size()));
+
+    DefensePlayBase::setAlignment(event, crease_defender_assignments);
+    DefensePlayBase::updatePassDefenderControlParams(pass_defender_assignments);
```

Comment:
Since this class extends `DefensePlayBase`, we don't need `DefensePlayBase::` in each of these funcitons